tests/ui/iterators/generator_returned_from_fn.rs

```diff
@@ -0,0 +1,65 @@
+#![feature(iter_macro, impl_trait_in_fn_trait_return)]
+// FIXME(iter_macro): make `yield` within it legal
+#![feature(coroutines)]
+
+use std::iter::iter;
+
+fn plain() -> impl Fn() -> impl Iterator<Item = u32> {
+    iter! {
```

Comment:
Oh this is interesting. I was expecting this case would return `impl IntoIterator`, not an `impl GenFn`-equivalent. I don't necessarily mind it though.

I do want to check that we're confident that we can create a blanket impl for `impl IntoIterator for GenFn {}` (or do something equivalent) so that the following snippet will just work once we add `GenFn` proper:

```rust
let iter = iter! {
    for x in 5..10 {
        yield x * 2;
    }
};

for x in iter { // <- calls `.into_iter` for you
    println!("{x}");
}
```

Reply:
I also would expect this to return `impl IntoIterator`. Maybe one day `impl Iterator`, but we can defer that discussion for later.

Reply:
I thought the recent direction has been to prefer functions as the more powerful version of `Into*`, so this seems in line with that.

That said, I think it's important that the output of the `iter!` can go directly into a `for` loop if there is no `||`.

Reply:
So I guess that's the surprising part here, that this example doesn't have `||`.